digdag-ui/console.jsx

```diff
@@ -1920,6 +1950,313 @@ class WorkflowRevisionPage extends React.Component {
   }
 }
 
+class FileEditor extends React.Component {
+  editor: ?CodeEditor
+
+  props:{
+    file: ?TarEntry;
+    projectArchive: ?ProjectArchive;
+  }
+
+  state:{
+    name: string;
+  }
+
+  constructor (props) {
+    super(props)
+    this.state = {
+      name: props.file ? props.file.name : 'new.dig'
+    }
+  }
+
+  getName () {
+    return this.state.name
+  }
+
+  getFileContents () {
+    var text
+    if (this.editor) {
+      text = this.editor.getValue()
+    } else {
+      text = this.props.file ? fileString(this.props.file.name, this.props.projectArchive) : ''
+    }
+
+    const buffer = new Buffer(text)
+    const ab = new ArrayBuffer(buffer.length);
+    const view = new Uint8Array(ab);
+    for (var i = 0; i < buffer.length; ++i) {
+      view[i] = buffer[i];
+    }
+    return ab;
+  }
+
+  render () {
+    const file = this.props.file
+    return (
+      <div>
+        <h3>
+          <input type="text" className="form-control" value={this.state.name} onChange={this.handleNameChange.bind(this)} />
+        </h3>
+        <pre>
+          <Measure>
+            { ({ width }) =>
+                <CodeEditor
+                  className='editor'
+                  language='yaml'  // TODO how to let ace guess language?
+                  value={file ? fileString(file.name, this.props.projectArchive) : ''}
+                  style={{ width }}
+                  ref={(vlaue) => { this.editor = vlaue }}
+                />
+            }
+          </Measure>
+        </pre>
+        <button className='btn btn-sm btn-error' onClick={this.props.onDelete}>Delete</button>
+      </div>
+    )
+  }
+
+  handleNameChange (event) {
+    this.setState({name: event.target.value});
+  }
+}
+
+class ProjectArchiveEditor extends React.Component {
+  props:{
+    projectArchive: ?ProjectArchive;
+  }
+
+  state:{
+    entries: [];
+  }
+
+  constructor (props) {
+    super(props)
+    this.state = {
+      entries: []
+    }
+  }
+
+  handleDelete (key) {
+    _.remove(this.state.entries, (file) => file.key === key)
+    this.setState({entries: this.state.entries})
+  }
+
+  handleAddFile () {
+    const key = uuid.v4()
+    this.state.entries.unshift({
+      newFile: Boolean(true),
+      key: key,
```

Comment:
Also is key really relevant (should we not generate it at render time rather than keeping it as state?